prov/rxm/src/rxm_cq.c

```diff
@@ -604,52 +604,56 @@ ssize_t rxm_sar_handle_segment(struct rxm_rx_buf *rx_buf)
 static ssize_t rxm_lmt_send_ack(struct rxm_rx_buf *rx_buf)
 {
 	struct rxm_tx_entry *tx_entry;
-	struct rxm_tx_buf *tx_buf;
 	ssize_t ret;
 
 	assert(rx_buf->conn);
 
-	tx_buf = rxm_tx_buf_get(rx_buf->ep, RXM_BUF_POOL_TX_ACK);
-	if (OFI_UNLIKELY(!tx_buf)) {
-		FI_WARN(&rxm_prov, FI_LOG_CQ, "TX queue full!\n");
+	rx_buf->recv_entry->rndv.tx_buf =
+		rxm_tx_buf_get(rx_buf->ep, RXM_BUF_POOL_TX_ACK);
+	if (OFI_UNLIKELY(!rx_buf->recv_entry->rndv.tx_buf)) {
+		FI_WARN(&rxm_prov, FI_LOG_CQ,
+			"Unable to allocate TX buffer\n");
 		return -FI_EAGAIN;
 	}
-	assert(tx_buf->pkt.ctrl_hdr.type == ofi_ctrl_ack);
-
-	tx_entry = rxm_tx_entry_get(rx_buf->conn->send_queue);
-	if (OFI_UNLIKELY(!tx_entry)) {
-		ret = -FI_EAGAIN;
-		goto err1;
-	}
+	assert(rx_buf->recv_entry->rndv.tx_buf->pkt.ctrl_hdr.type == ofi_ctrl_ack);
 
 	RXM_LOG_STATE(FI_LOG_CQ, rx_buf->pkt, RXM_LMT_READ, RXM_LMT_ACK_SENT);
 	rx_buf->hdr.state = RXM_LMT_ACK_SENT;
 
-	tx_entry->conn = rx_buf->conn;
-	tx_entry->state = rx_buf->hdr.state;
-	tx_entry->context = rx_buf;
-	tx_entry->tx_buf = tx_buf;
-
-	tx_buf->pkt.ctrl_hdr.conn_id 	= rx_buf->conn->handle.remote_key;
-	tx_buf->pkt.ctrl_hdr.msg_id 	= rx_buf->pkt.ctrl_hdr.msg_id;
+	rx_buf->recv_entry->rndv.tx_buf->pkt.ctrl_hdr.conn_id =
+		rx_buf->conn->handle.remote_key;
+	rx_buf->recv_entry->rndv.tx_buf->pkt.ctrl_hdr.msg_id =
+		rx_buf->pkt.ctrl_hdr.msg_id;
 
-	ret = fi_send(rx_buf->conn->msg_ep, &tx_buf->pkt, sizeof(tx_buf->pkt),
-		      tx_buf->hdr.desc, 0, tx_entry);
+	ret = fi_send(rx_buf->conn->msg_ep, &rx_buf->recv_entry->rndv.tx_buf->pkt,
+		      sizeof(rx_buf->recv_entry->rndv.tx_buf->pkt),
+		      rx_buf->recv_entry->rndv.tx_buf->hdr.desc, 0, rx_buf);
```

Comment:
Why is allocating tx_entry being delayed until deferred ACK? If we pass rx_buf as context how are we tracking ACK send completions?

Reply:
If there is no space in TX queue, `rxm_tx_entry_get` fails. So, this leads that we can't send the ACK.
It's ok to track the CQ entry ACK for rx_buf. The rx_buf is with `RXM_LMT_ACK_SENT` state in this case.